File: backend/src/main/java/com/technicalescaperoom/backend/controller/admin/AdminContentController.java
package com.technicalescaperoom.backend.controller.admin;

import com.technicalescaperoom.backend.config.security.AdminPrincipal;
import com.technicalescaperoom.backend.dto.admin.*;
import com.technicalescaperoom.backend.service.admin.AdminContentService;
import com.technicalescaperoom.backend.service.admin.EventContentValidationService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/admin/events")
@RequiredArgsConstructor
public class AdminContentController {

    private final AdminContentService adminContentService;
    private final EventContentValidationService eventContentValidationService;

    @GetMapping("/{eventId}/content")
    public ResponseEntity<Map<String, Object>> getAllContent(@PathVariable Long eventId) {
        Map<String, Object> content = adminContentService.getAllContentForEvent(eventId);
        return ResponseEntity.ok(content);
    }

    @GetMapping("/{eventId}/validation")
    public ResponseEntity<EventReadinessDto> validateEventReadiness(@PathVariable Long eventId) {
        EventReadinessDto readiness = eventContentValidationService.validateEventReadiness(eventId);
        return ResponseEntity.ok(readiness);
    }

    @PutMapping("/{eventId}/levels/{levelNumber}/questions")
    public ResponseEntity<QuestionConfigDto> saveQuestion(
            @PathVariable Long eventId,
            @PathVariable Integer levelNumber,
            @Valid @RequestBody QuestionConfigDto dto,
            @AuthenticationPrincipal AdminPrincipal principal) {
        QuestionConfigDto response = adminContentService.saveQuestionConfig(principal, eventId, levelNumber, dto);
        return ResponseEntity.ok(response);
    }

    @PutMapping("/{eventId}/levels/{levelNumber}/hint")
    public ResponseEntity<HintConfigDto> saveHint(
            @PathVariable Long eventId,
            @PathVariable Integer levelNumber,
            @Valid @RequestBody HintConfigDto dto,
            @AuthenticationPrincipal AdminPrincipal principal) {
        HintConfigDto response = adminContentService.saveHintConfig(principal, eventId, levelNumber, dto);
        return ResponseEntity.ok(response);
    }

    @PostMapping("/{eventId}/test-answer")
    public ResponseEntity<AnswerTestResponseDto> testAnswer(
            @PathVariable Long eventId,
            @Valid @RequestBody AnswerTestRequestDto request,
            @AuthenticationPrincipal AdminPrincipal principal) {
        AnswerTestResponseDto response = adminContentService.testAnswer(principal, eventId, request);
        return ResponseEntity.ok(response);
    }

    @GetMapping("/{eventId}/preview/player")
    public ResponseEntity<PlayerSafePreviewDto> getPlayerSafePreview(
            @PathVariable Long eventId,
            @RequestParam Integer levelNumber,
            @RequestParam Integer playerNumber) {
        PlayerSafePreviewDto preview = adminContentService.getPlayerSafePreview(eventId, levelNumber, playerNumber);
        return ResponseEntity.ok(preview);
    }
}

File: backend/controllers/admin/adminContentController.js
const adminContentService = require('../../services/admin/adminContentService');
const eventContentValidationService = require('../../services/admin/eventContentValidationService');

class AdminContentController {
  async getAllContent(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      const content = await adminContentService.getAllContentForEvent(eventId);
      res.json(content);
    } catch (err) {
      next(err);
    }
  }

  async validateEventReadiness(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      const readiness = await eventContentValidationService.validateEventReadiness(eventId);
      res.json(readiness);
    } catch (err) {
      next(err);
    }
  }

  async saveQuestion(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      const levelNumber = parseInt(req.params.levelNumber, 10);
      const response = await adminContentService.saveQuestionConfig(req.admin, eventId, levelNumber, req.body);
      res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async saveHint(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      const levelNumber = parseInt(req.params.levelNumber, 10);
      const response = await adminContentService.saveHintConfig(req.admin, eventId, levelNumber, req.body);
      res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async testAnswer(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      const response = await adminContentService.testAnswer(req.admin, eventId, req.body);
      res.json(response);
    } catch (err) {
      next(err);
    }
  }

  async getPlayerSafePreview(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      const levelNumber = parseInt(req.query.levelNumber, 10);
      const playerNumber = parseInt(req.query.playerNumber, 10);
      const preview = await adminContentService.getPlayerSafePreview(eventId, levelNumber, playerNumber);
      res.json(preview);
    } catch (err) {
      next(err);
    }
  }
}

module.exports = new AdminContentController();
